File: leetcode/Python/0040-m-combination-sum-ii.py
"""
DFS
"""
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        self.helper(sorted(candidates), [], 0, target, res)
        return res

    def helper(self, candidates, path, start, target, res):
        if target == 0:
            res.append(path[:])
            return

        for i in range(start, len(candidates)):
            if candidates[i] > target:
                return
            if i > start and candidates[i] == candidates[i - 1]:
                continue
            path.append(candidates[i])
            self.helper(candidates, path, i + 1, target - candidates[i], res)
            path.pop()

File: leetcode/C++/0040-m-combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        vector<int> path;
        dfs(candidates, target, 0, path, res);
        return res;
    }
    
    void dfs(vector<int>& candidates, int target, int start, 
             vector<int>& path, vector<vector<int>>& res) {
        if (target == 0)
            res.push_back(path);
        for (int i = start; i < candidates.size(); i++) {
            if (candidates[i] > target)
                return;
            if (i > start && candidates[i] == candidates[i - 1])
                continue;
            path.push_back(candidates[i]);
            dfs(candidates, target - candidates[i], i + 1, path, res);
            path.pop_back();
        }
    }
};